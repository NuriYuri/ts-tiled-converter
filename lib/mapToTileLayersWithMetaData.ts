import { layerToTiles } from './layerToTiles';
import { TiledXMLGroup, TiledXMLLayer, TiledXMLMap, getXMLProperties } from './tiledXML/objects';
import type { LayerWithMetaData, TileMetaData } from './types';
import { throwIfError, toError } from './util';

let cache: Map<number, TileMetaData | null>;

export const mapToTileLayersWithMetaData = ({ map }: TiledXMLMap): LayerWithMetaData[] | Error => {
  try {
    cache = new Map<number, TileMetaData | null>();
    return filterAndFlatten(map, 0).map(mapLayerToTileLayers);
  } catch (error) {
    return toError(error);
  }
};

type AnyMapOrGroupChild = (TiledXMLMap['map'] | TiledXMLGroup['group'])[number];
type LayerWithZIndex = { z: number; name: string; entity: TiledXMLLayer };

const filterLayerOrGroup = (entity: AnyMapOrGroupChild): entity is TiledXMLLayer | TiledXMLGroup => 'layer' in entity || 'group' in entity;

const filterAndFlatten = (e: AnyMapOrGroupChild[], groupZIndex: number) => e.filter(filterLayerOrGroup).flatMap(mapGroupOrLayerToLayers(groupZIndex));

const flattenGroup = (group: TiledXMLGroup, groupZIndex: number): LayerWithZIndex[] => {
  const { name } = getXMLProperties(group);
  return filterAndFlatten(group.group, name.startsWith('z=') ? Number(name.split('=')[1]) - 1 : groupZIndex);
};

const mapGroupOrLayerToLayers = (groupZIndex: number) => (entity: TiledXMLLayer | TiledXMLGroup) => {
  if ('group' in entity) return flattenGroup(entity, groupZIndex);

  const { name } = getXMLProperties(entity);
  const regMatch = name.match(/([1-6])$/);
  return { z: regMatch ? Number(regMatch[0]) - 1 : groupZIndex, name, entity };
};

const mapLayerToTileLayers = ({ z, name, entity }: LayerWithZIndex) => ({ z, name, layer: throwIfError(layerToTiles(entity, cache)) });
