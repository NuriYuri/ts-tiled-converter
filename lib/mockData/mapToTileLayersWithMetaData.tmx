<?xml version="1.0" encoding="UTF-8"?>
<map version="1.10" tiledversion="1.10.1" orientation="orthogonal" renderorder="right-down" width="2" height="2" tilewidth="32" tileheight="32" infinite="0" nextlayerid="23" nextobjectid="1">
   <tileset firstgid="1" source="../../../../comissions2/demo/Tilesets/Atori_borders.tsx"/>
   <layer id="9996" name="orphan1" width="2" height="2" visible="0">
      <data encoding="csv">
1,1,
1,0
      </data>
   </layer>
   <layer id="9997" name="orphan2" width="2" height="2" visible="0">
      <data encoding="csv">
2,2,
0,2
      </data>
   </layer>
   <layer id="6" name="orphan1" width="2" height="2" visible="1">
      <data encoding="csv">
1,1,
1,0
      </data>
   </layer>
   <layer id="7" name="orphan2" width="2" height="2" visible="1">
      <data encoding="csv">
2,2,
0,2
      </data>
   </layer>
   <group id="998" name="z=1" visible="0">
      <layer id="999" name="ground" width="2" height="2">
         <data encoding="csv">
0,3,
0,3
         </data>
      </layer>
      <layer id="9910" name="badlySorted2" width="2" height="2">
         <data encoding="csv">
4,0,
0,4
         </data>
      </layer>
   </group>
   <group id="8" name="z=1" visible="1">
      <layer id="9" name="ground" width="2" height="2">
         <data encoding="csv">
0,3,
0,3
         </data>
      </layer>
      <layer id="10" name="badlySorted2" width="2" height="2">
         <data encoding="csv">
4,0,
0,4
         </data>
      </layer>
   </group>
   <group id="99911" name="z=2" visible="0">
      <layer id="12" name="badlySorted1" width="2" height="2">
         <data encoding="csv">
201,201,
0,201
         </data>
      </layer>
      <layer id="99913" name="trees" width="2" height="2">
         <data encoding="csv">
64,64,
0,0
         </data>
      </layer>
      <group id="99915" name="Sub Folder">
         <layer id="99916" name="trees_top" width="2" height="2">
            <data encoding="csv">
0,25,
0,0
            </data>
         </layer>
      </group>
   </group>
   <group id="11" name="z=2" visible="1">
      <layer id="12" name="badlySorted1" width="2" height="2">
         <data encoding="csv">
201,201,
0,201
         </data>
      </layer>
      <layer id="13" name="trees" width="2" height="2">
         <data encoding="csv">
64,64,
0,0
         </data>
      </layer>
      <group id="15" name="Sub Folder">
         <layer id="16" name="trees_top" width="2" height="2">
            <data encoding="csv">
0,25,
0,0
            </data>
         </layer>
      </group>
   </group>
   <group id="17" name="z=3">
      <group id="19" name="before">
         <layer id="21" name="inBefore" width="2" height="2">
            <data encoding="csv">
0,0,
0,0
            </data>
         </layer>
      </group>
      <layer id="20" name="middle" width="2" height="2">
         <data encoding="csv">
0,0,
0,0
         </data>
      </layer>
      <group id="18" name="after">
         <layer id="22" name="inAfter" width="2" height="2">
            <data encoding="csv">
0,0,
0,0
            </data>
         </layer>
      </group>
   </group>
   <group id="2" name="system" visible="0">
      <layer id="5" name="passages" width="2" height="2" visible="0">
         <data encoding="csv">
0,7,
7,0
         </data>
      </layer>
      <layer id="1" name="systemtags" width="2" height="2">
         <data encoding="csv">
85,0,
0,85
         </data>
      </layer>
      <layer id="3" name="systemtags_bridge1" width="2" height="2">
         <data encoding="csv">
65,65,
0,65
         </data>
      </layer>
      <layer id="4" name="systemtags_bridge2" width="2" height="2">
         <data encoding="csv">
0,66,
66,66
         </data>
      </layer>
   </group>
</map>
