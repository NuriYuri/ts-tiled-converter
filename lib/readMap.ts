import { XMLParser } from 'fast-xml-parser';
import fs from 'fs';
import crypto from 'crypto';

const NUMERIC_ATTRIBUTES = [
  'compressionlevel',
  'width',
  'height',
  'tilewidth',
  'tileheight',
  'infinite',
  'nextlayerid',
  'nextobjectid',
  'firstgid',
  'id',
];

const ARRAY_TAGS = ['tilesets', 'groups', 'layers'];

const TAG_NAME_DICTIONARY = {
  tileset: 'tilesets',
  group: 'groups',
  layer: 'layers',
} as Record<string, string>;

const ATTRIBUTE_NAME_DICTIONARY = {
  tilewidth: 'tileWidth',
  tileheight: 'tileHeight',
  nextlayerid: 'nextLayerId',
  nextobjectid: 'nextObjectId',
  firstgid: 'firstGlobalId',
  compressionlevel: 'compressionLevel',
} as Record<string, string>;

export type TMXData = {
  tilesets: { firstGlobalId: number; source: string }[];
  groups: { id: number; name: string }[];
  layers: {
    id: number;
    name: string;
    width: number;
    height: number;
    data: {
      encoding: 'base64' | 'csv';
      compression?: 'gzip' | 'zlib' | 'zstd';
      rawData: string;
    };
  };
  version: string;
  tiledversion: string;
  orientation: 'orthogonal';
  renderorder: 'right-down';
  compressionLevel: number;
  width: number;
  height: number;
  tileWidth: number;
  tileHeight: number;
  infinite: number;
  nextLayerId: number;
  nextObjectId: number;
  sha1: string;
};

export const readMap = (path: string) => {
  try {
    const parser = new XMLParser({
      ignoreAttributes: false,
      attributeNamePrefix: '',
      ignorePiTags: true,
      textNodeName: 'rawData',
      transformTagName: (name) => TAG_NAME_DICTIONARY[name] || name,
      transformAttributeName: (name) => ATTRIBUTE_NAME_DICTIONARY[name] || name,
      attributeValueProcessor: (attrName, value) => (NUMERIC_ATTRIBUTES.includes(attrName) ? Number(value) : undefined),
      isArray: (name) => ARRAY_TAGS.includes(name),
    });
    const fileData = fs.readFileSync(path);
    const result = parser.parse(fileData);
    return { groups: [], layers: [], tilesets: [], ...result.map, sha1: crypto.createHash('sha1').update(fileData).digest('hex') } as TMXData;
  } catch (error) {
    if (error instanceof Error) return error;
    return new Error(`Unknown error: ${JSON.stringify(error)}`);
  }
};
