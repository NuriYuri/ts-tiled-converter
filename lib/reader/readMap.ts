import fs from 'fs';
import crypto from 'crypto';
import { decodeXML } from '../tiledXML/decodeXML';
import type { TiledXMLMap } from '../tiledXML/objects';

export const readMap = (path: string) => {
  try {
    const fileData = fs.readFileSync(path);
    const map = decodeXML(fileData)[0] as TiledXMLMap;
    return {
      map,
      sha1: crypto.createHash('sha1').update(fileData).digest('hex'),
    };
  } catch (error) {
    if (error instanceof Error) return error;
    return new Error(`Unknown error: ${JSON.stringify(error)}`);
  }
};
