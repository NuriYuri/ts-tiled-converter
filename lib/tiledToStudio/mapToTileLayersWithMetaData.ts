import { layerToTiles } from './layerToTiles';
import { SPECIAL_LAYER_TILESETS } from './splitSpecialAndRegularLayers';
import { TiledXMLGroup, TiledXMLLayer, TiledXMLMap, getXMLProperties } from '../tiledXML/objects';
import type { LayerWithMetaData, TileMetaData } from '../types';
import { throwIfError, toError } from '../util';

let cache: Map<number, TileMetaData | null>;

export const mapToTileLayersWithMetaData = ({ map }: TiledXMLMap): LayerWithMetaData[] | Error => {
  try {
    cache = new Map<number, TileMetaData | null>();
    return filterAndFlatten(map, 0).map(mapLayerToTileLayers);
  } catch (error) {
    return toError(error);
  }
};

type AnyMapOrGroupChild = (TiledXMLMap['map'] | TiledXMLGroup['group'])[number];
type LayerWithZIndex = { z: number; name: string; entity: TiledXMLLayer };

const isVisibleLayer = (entity: AnyMapOrGroupChild): entity is TiledXMLLayer =>
  'layer' in entity && (getXMLProperties(entity).visible != 0 || getXMLProperties(entity).name in SPECIAL_LAYER_TILESETS);

const isVisibleGroup = (entity: AnyMapOrGroupChild): entity is TiledXMLGroup =>
  'group' in entity && (getXMLProperties(entity).visible != 0 || getXMLProperties(entity).name.startsWith('system'));

const filterLayerOrGroup = (entity: AnyMapOrGroupChild): entity is TiledXMLLayer | TiledXMLGroup => isVisibleLayer(entity) || isVisibleGroup(entity);

const filterAndFlatten = (e: AnyMapOrGroupChild[], groupZIndex: number) => e.filter(filterLayerOrGroup).flatMap(mapGroupOrLayerToLayers(groupZIndex));

const flattenGroup = (group: TiledXMLGroup, groupZIndex: number): LayerWithZIndex[] => {
  const { name } = getXMLProperties(group);
  return filterAndFlatten(group.group, name.startsWith('z=') ? Number(name.split('=')[1]) - 1 : groupZIndex);
};

const mapGroupOrLayerToLayers = (groupZIndex: number) => (entity: TiledXMLLayer | TiledXMLGroup) => {
  if ('group' in entity) return flattenGroup(entity, groupZIndex);

  const { name } = getXMLProperties(entity);
  const regMatch = name.match(/([1-6])$/);
  return { z: regMatch ? Number(regMatch[0]) - 1 : groupZIndex, name, entity };
};

const mapLayerToTileLayers = ({ z, name, entity }: LayerWithZIndex) => ({ z, name, layer: throwIfError(layerToTiles(entity, cache)) });
