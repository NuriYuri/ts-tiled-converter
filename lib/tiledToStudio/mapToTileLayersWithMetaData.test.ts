import { readMap } from '../reader/readMap';
import { mapToTileLayersWithMetaData } from './mapToTileLayersWithMetaData';

describe('mapToTileLayersWithMetaData', () => {
  it('maps data properly', () => {
    const map = readMap('lib/mockData/mapToTileLayersWithMetaData.tmx');
    if (map instanceof Error) throw map;

    expect(mapToTileLayersWithMetaData(map.map)).toMatchSnapshot();
  });
});
