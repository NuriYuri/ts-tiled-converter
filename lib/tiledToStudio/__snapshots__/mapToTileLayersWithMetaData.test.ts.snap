// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapToTileLayersWithMetaData maps data properly 1`] = `
[
  {
    "layer": [
      {
        "globalId": 1,
        "transformId": 0,
      },
      {
        "globalId": 1,
        "transformId": 0,
      },
      {
        "globalId": 1,
        "transformId": 0,
      },
      null,
    ],
    "name": "orphan1",
    "z": 0,
  },
  {
    "layer": [
      {
        "globalId": 2,
        "transformId": 0,
      },
      {
        "globalId": 2,
        "transformId": 0,
      },
      null,
      {
        "globalId": 2,
        "transformId": 0,
      },
    ],
    "name": "orphan2",
    "z": 1,
  },
  {
    "layer": [
      null,
      {
        "globalId": 3,
        "transformId": 0,
      },
      null,
      {
        "globalId": 3,
        "transformId": 0,
      },
    ],
    "name": "ground",
    "z": 0,
  },
  {
    "layer": [
      {
        "globalId": 4,
        "transformId": 0,
      },
      null,
      null,
      {
        "globalId": 4,
        "transformId": 0,
      },
    ],
    "name": "badlySorted2",
    "z": 1,
  },
  {
    "layer": [
      {
        "globalId": 201,
        "transformId": 0,
      },
      {
        "globalId": 201,
        "transformId": 0,
      },
      null,
      {
        "globalId": 201,
        "transformId": 0,
      },
    ],
    "name": "badlySorted1",
    "z": 0,
  },
  {
    "layer": [
      {
        "globalId": 64,
        "transformId": 0,
      },
      {
        "globalId": 64,
        "transformId": 0,
      },
      null,
      null,
    ],
    "name": "trees",
    "z": 1,
  },
  {
    "layer": [
      null,
      {
        "globalId": 25,
        "transformId": 0,
      },
      null,
      null,
    ],
    "name": "trees_top",
    "z": 1,
  },
  {
    "layer": [
      null,
      null,
      null,
      null,
    ],
    "name": "inBefore",
    "z": 2,
  },
  {
    "layer": [
      null,
      null,
      null,
      null,
    ],
    "name": "middle",
    "z": 2,
  },
  {
    "layer": [
      null,
      null,
      null,
      null,
    ],
    "name": "inAfter",
    "z": 2,
  },
  {
    "layer": [
      null,
      {
        "globalId": 7,
        "transformId": 0,
      },
      {
        "globalId": 7,
        "transformId": 0,
      },
      null,
    ],
    "name": "passages",
    "z": 0,
  },
  {
    "layer": [
      {
        "globalId": 85,
        "transformId": 0,
      },
      null,
      null,
      {
        "globalId": 85,
        "transformId": 0,
      },
    ],
    "name": "systemtags",
    "z": 0,
  },
  {
    "layer": [
      {
        "globalId": 65,
        "transformId": 0,
      },
      {
        "globalId": 65,
        "transformId": 0,
      },
      null,
      {
        "globalId": 65,
        "transformId": 0,
      },
    ],
    "name": "systemtags_bridge1",
    "z": 0,
  },
  {
    "layer": [
      null,
      {
        "globalId": 66,
        "transformId": 0,
      },
      {
        "globalId": 66,
        "transformId": 0,
      },
      {
        "globalId": 66,
        "transformId": 0,
      },
    ],
    "name": "systemtags_bridge2",
    "z": 1,
  },
]
`;
